File: bench/cc02_mutex_contention.zig
// cc02_mutex_contention.zig — 并发基准：互斥锁争用（Zig -O2 对照）
// M=4 线程各在共享互斥锁下对 counter 自增 N_ITERS=1e7 次，终值 = M*N。
const std = @import("std");

const N_THREADS: usize = 4;
const N_ITERS: i64 = 10_000_000;

fn worker(mtx: *std.atomic.Mutex, counter: *i64, iters: i64) void {
    var i: i64 = 0;
    while (i < iters) : (i += 1) {
        while (!mtx.tryLock()) {}
        counter.* += 1;
        mtx.unlock();
    }
}

pub fn main() !void {
    var mtx: std.atomic.Mutex = .unlocked;
    var counter: i64 = 0;
    var threads: [N_THREADS]std.Thread = undefined;
    var t: usize = 0;
    while (t < N_THREADS) : (t += 1) {
        threads[t] = try std.Thread.spawn(.{}, worker, .{ &mtx, &counter, N_ITERS });
    }
    t = 0;
    while (t < N_THREADS) : (t += 1) {
        threads[t].join();
    }
    asm volatile ("" : [counter] "+r" (counter) : : .{ .memory = true });
    std.process.exit(@intCast(counter & 255));
}

File: bench/cc02_mutex_contention.c
/**
 * cc02_mutex_contention.c — 并发基准：互斥锁争用（C -O2 -pthread 对照）
 *
 * M=4 线程各在共享互斥锁保护下对 counter 做 N_ITERS=10000000 次自增，
 * 终值 = M*N = 40000000。测量 pthread_mutex 在高争用下的吞吐。
 *
 * 编译：cc -O2 -pthread bench/cc02_mutex_contention.c -o /tmp/test_cc02
 */
#include <pthread.h>
#include <stdint.h>
#include <stdio.h>

enum { N_THREADS = 4 };
#define N_ITERS 10000000LL

typedef struct {
    pthread_mutex_t *mtx;
    int64_t *counter;
    int64_t iters;
} worker_arg_t;

/** worker：在互斥锁保护下对 *counter 自增 iters 次。 */
static void *worker(void *arg) {
    worker_arg_t *a = (worker_arg_t *)arg;
    int64_t i = 0;
    while (i < a->iters) {
        pthread_mutex_lock(a->mtx);
        *a->counter += 1;
        pthread_mutex_unlock(a->mtx);
        i = i + 1;
    }
    return NULL;
}

int main(void) {
    pthread_mutex_t mtx;
    if (pthread_mutex_init(&mtx, NULL) != 0) {
        fprintf(stderr, "cc02: mutex_init failed\n");
        return 1;
    }
    int64_t counter = 0;
    pthread_t threads[N_THREADS];
    worker_arg_t args[N_THREADS];
    int32_t t = 0;
    while (t < N_THREADS) {
        args[t].mtx = &mtx;
        args[t].counter = &counter;
        args[t].iters = N_ITERS;
        if (pthread_create(&threads[t], NULL, worker, &args[t]) != 0) {
            fprintf(stderr, "cc02: pthread_create %d failed\n", (int)t);
            return 2;
        }
        t = t + 1;
    }
    t = 0;
    while (t < N_THREADS) {
        if (pthread_join(threads[t], NULL) != 0) {
            fprintf(stderr, "cc02: pthread_join %d failed\n", (int)t);
            return 3;
        }
        t = t + 1;
    }
    pthread_mutex_destroy(&mtx);
    /* 防止 counter 被折叠。 */
    __asm__ volatile("" : "+r"(counter) : : "memory");
    return (int)(counter & 255);
}
